compiler/rustc_trait_selection/src/solve/fulfill.rs

```diff
@@ -314,10 +314,36 @@ impl<'tcx> ProofTreeVisitor<'tcx> for BestObligation<'tcx> {
     }
 
     fn visit_goal(&mut self, goal: &super::inspect::InspectGoal<'_, 'tcx>) -> Self::Result {
-        // FIXME: Throw out candidates that have no failing WC and >0 failing misc goal.
+        // Filter out the candidates that aren't either error or ambiguous (depending
+        // on what we are looking for).
+        let mut candidates = goal
+            .candidates()
+            .into_iter()
+            .filter(|candidate| match self.consider_ambiguities {
+                true => matches!(candidate.result(), Ok(Certainty::Maybe(_))),
+                false => matches!(candidate.result(), Err(_)),
+            })
+            .collect::<Vec<_>>();
+
+        // Throw out candidates that have no failing WC (or higher-ranked obligations,
+        // for which there should only be one candidate anyways -- but I digress).
         // This most likely means that the goal just didn't unify at all, e.g. a param
         // candidate with an alias in it.
-        let candidates = goal.candidates();
+        if candidates.len() > 1 {
+            candidates.retain(|candidate| {
+                goal.infcx().probe(|_| {
```

Comment:
Kind of a shame we have to do this in a probe, but if we don't, then the instantiated nested goals from different candidates will affect eachother, causing weirdness and ICEs.

Luckily, we do this only when there's >1 candidate.

We could, perhaps, first filter out candidates with *no* `GoalSource::ImplWhereBound | GoalSource::InstantiateHigherRanked` without even instantiating them, and only instantiate them if necessary.

Also, I guess we should have both the source *and* the certainty in the proof tree without instantiating -- we just need to expose it, perhaps?